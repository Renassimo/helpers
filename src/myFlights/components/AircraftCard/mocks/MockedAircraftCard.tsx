const MockedAircraftCard = jest.fn(({ chosen }) => (
  <div>MockedAircraftCard - {chosen && 'chosen'}</div>
));

export default MockedAircraftCard;
