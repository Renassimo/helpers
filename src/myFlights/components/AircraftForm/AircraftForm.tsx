import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import BaseAircraftForm from '@/avia/components/BaseAircraftForm';

const AircraftForm = ({ registration }: { registration?: string | null }) => {
  const { aircraftsResult } = useMyFlightsContext();

  return (
    <BaseAircraftForm
      registration={registration}
      aircraftsResult={aircraftsResult}
    />
  );
};

export default AircraftForm;
