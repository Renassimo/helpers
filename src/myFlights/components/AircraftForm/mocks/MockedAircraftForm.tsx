const MockedAircraftForm = jest.fn(({ registration }) => (
  <div>MockedAircraftForm - {registration}</div>
));

export default MockedAircraftForm;
