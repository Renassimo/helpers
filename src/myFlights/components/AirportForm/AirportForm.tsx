import { useState, KeyboardEvent, useEffect } from 'react';

import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import LabelIcon from '@mui/icons-material/Label';
import TitleIcon from '@mui/icons-material/Title';
import MyLocationIcon from '@mui/icons-material/MyLocation';

import useInputValue from '@/common/hooks/useInputValue';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import AirportCard from '@/myFlights/components/AirportCard';

import AviaInput from '@/avia/components/AviaInput';

import useAlerts from '@/common/hooks/alerts/useAlerts';
import geoLocation from '@/common/utils/geoLocation';

const AirportForm = ({ title }: { title: 'Origin' | 'Destination' }) => {
  const { createErrorAlert } = useAlerts();

  const key = title === 'Origin' ? 'originsResult' : 'destinationsResult';
  const {
    [key]: {
      airports,
      chosenAirport,
      retreiveAirports,
      chooseAirport,
      clearChosenAirport,
      loading,
    },
  } = useMyFlightsContext();

  const [airportValue, setAirportValue] = useInputValue<string>('');
  const [by, setBy] = useState<'code' | 'text' | 'location'>('code');
  const [locationLoading, setLocationLoading] = useState(false);

  useEffect(() => {
    if (airportValue.length > 4 && by === 'code') setBy('text');
  }, [airportValue]);

  const searchByLocation = async () => {
    setLocationLoading(true);
    geoLocation(
      async (args) =>
        await retreiveAirports({
          lat: String(args.coords.latitude),
          lon: String(args.coords.longitude),
        }),
      ({ message }) => createErrorAlert(message),
      createErrorAlert,
      () => {
        setAirportValue('');
        setBy('location');
      }
    );
    setLocationLoading(false);
  };

  const searchByCode = async () => {
    setBy('code');
    await retreiveAirports({
      code: airportValue,
    });
  };

  const searchByText = async () => {
    setBy('text');
    await retreiveAirports({
      text: airportValue,
    });
  };

  const onEnter = (event: KeyboardEvent) => {
    if (event.key === 'Enter') {
      if (airportValue.length > 4) {
        searchByText();
      } else if (airportValue.length > 2) {
        searchByCode();
      }
    }
  };

  const inputDisabled = loading || locationLoading;
  const textingDisabled = inputDisabled || airportValue.length < 3;

  return (
    <Box mb={1}>
      {!chosenAirport && (
        <AviaInput
          title={title}
          placeholder={`${title} IATA or ICAO code`}
          label={title}
          value={airportValue}
          setValue={setAirportValue}
          disabled={inputDisabled}
          onKeyDown={onEnter}
          adornments={
            <>
              <IconButton
                size="small"
                color={by === 'code' ? 'primary' : undefined}
                onClick={searchByCode}
                disabled={textingDisabled}
                aria-label="By Code"
              >
                <LabelIcon />
              </IconButton>
              <IconButton
                size="small"
                color={by === 'text' ? 'primary' : undefined}
                onClick={searchByText}
                disabled={textingDisabled}
                aria-label="By Text"
              >
                <TitleIcon />
              </IconButton>
              <IconButton
                size="small"
                color={by === 'location' ? 'primary' : undefined}
                onClick={searchByLocation}
                disabled={inputDisabled}
                aria-label="By Location"
              >
                <MyLocationIcon />
              </IconButton>
            </>
          }
        />
      )}
      {airports && !chosenAirport && (
        <Box display="flex" overflow="auto">
          {airports.map((airport, index) => (
            <Box
              key={airport.id + index}
              mr={1}
              ml={index === 0 ? 1 : undefined}
            >
              <AirportCard
                data={airport}
                onClick={() => chooseAirport(airport.id)}
              />
            </Box>
          ))}
        </Box>
      )}
      {chosenAirport && (
        <AirportCard
          data={chosenAirport}
          onClick={clearChosenAirport}
          title={title}
          chosen
        />
      )}
    </Box>
  );
};

export default AirportForm;
