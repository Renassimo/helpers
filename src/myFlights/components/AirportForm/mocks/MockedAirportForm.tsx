const MockedAirportForm = jest.fn(({ title }) => (
  <div>MockedAirportForm - {title}</div>
));

export default MockedAirportForm;
