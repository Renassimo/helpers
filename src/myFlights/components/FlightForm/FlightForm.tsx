import { KeyboardEvent } from 'react';

import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import SearchIcon from '@mui/icons-material/Search';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import useInputValue from '@/common/hooks/useInputValue';

import FlightCard from '@/myFlights/components/FlightCard';
import AviaInput from '@/avia/components/AviaInput';
import DatePicker from '@/common/components/DateInput';

const FlightForm = () => {
  const {
    flightsResult: {
      flights,
      chosenFlight,
      retreiveFlights,
      chooseFlight,
      clearChosenFlight,
      loading,
    },
  } = useMyFlightsContext();

  const [flightValue, setFlightValue] = useInputValue<string>('');
  const [flightDate, setFlightDate] = useInputValue<string | null>(null);

  const onSubmit = async () => {
    await retreiveFlights(flightValue, flightDate);
  };

  const lengthOk = flightValue.length > 2;

  const onEnter = (event: KeyboardEvent) => {
    if (event.key === 'Enter' && lengthOk) onSubmit();
  };

  return (
    <Box>
      {!chosenFlight && (
        <>
          <AviaInput
            title="Flight number"
            value={flightValue}
            setValue={setFlightValue}
            disabled={loading}
            onKeyDown={onEnter}
            adornments={
              <IconButton
                size="small"
                onClick={onSubmit}
                disabled={loading || !lengthOk}
                aria-label="Search"
              >
                <SearchIcon />
              </IconButton>
            }
          />
          <Box mt={2}>
            <DatePicker
              value={flightDate}
              setValue={(newValue) => setFlightDate(newValue)}
              placeholder="Flight date"
              disabled={loading}
              fullWidth
            />
          </Box>
        </>
      )}
      {flights && !chosenFlight && (
        <Box display="flex" overflow="auto">
          {flights.map((flight, index) => (
            <Box
              key={flight.id + index}
              mr={1}
              ml={index === 0 ? 1 : undefined}
            >
              <FlightCard
                data={flight}
                onClick={() => chooseFlight(flight.id)}
              />
            </Box>
          ))}
        </Box>
      )}
      {chosenFlight && (
        <FlightCard data={chosenFlight} onClick={clearChosenFlight} chosen />
      )}
    </Box>
  );
};

export default FlightForm;
