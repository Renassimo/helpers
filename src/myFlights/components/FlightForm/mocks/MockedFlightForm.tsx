const MockedFlightForm = jest.fn(() => <div>MockedFlightForm</div>);

export default MockedFlightForm;
