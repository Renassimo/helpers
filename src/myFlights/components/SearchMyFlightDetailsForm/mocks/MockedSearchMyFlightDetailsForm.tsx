const MockedSearchMyFlightDetailsForm = jest.fn(() => (
  <div>MockedSearchMyFlightDetailsForm</div>
));

export default MockedSearchMyFlightDetailsForm;
