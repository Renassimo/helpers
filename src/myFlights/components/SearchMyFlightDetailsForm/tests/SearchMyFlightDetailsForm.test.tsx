import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import AircraftForm from '@/myFlights/components/AircraftForm';
import AirportForm from '@/myFlights/components/AirportForm';
import FlightForm from '@/myFlights/components/FlightForm';

import useMyFlightsContext from '@/myFlights/contexts/hooks/useMyFlightsContext';

import MockedAircraftForm from '@/myFlights/components/AircraftForm/mocks';
import MockedAirportForm from '@/myFlights/components/AirportForm/mocks';
import MockedFlightForm from '@/myFlights/components/FlightForm/mocks';

import { mockedDeserializedFlights } from '@/avia/types/avia/mocks';

import SearchMyFlightDetailsForm from '../SearchMyFlightDetailsForm';

jest.mock('@/myFlights/components/AircraftForm');
jest.mock('@/myFlights/components/AirportForm');
jest.mock('@/myFlights/components/FlightForm');
jest.mock('@/myFlights/contexts/hooks/useMyFlightsContext');

describe('SearchMyFlightDetailsForm', () => {
  beforeEach(() => {
    (AircraftForm as unknown as jest.Mock).mockImplementation(
      MockedAircraftForm
    );
    (AirportForm as unknown as jest.Mock).mockImplementation(MockedAirportForm);
    (FlightForm as unknown as jest.Mock).mockImplementation(MockedFlightForm);
  });

  afterEach(() => {
    jest.fn();
  });

  test('renders successfully', () => {
    // Arange
    (useMyFlightsContext as unknown as jest.Mock).mockImplementation(
      jest.fn(() => ({ flightsResult: {} }))
    );
    // Act
    const { baseElement } = renderWithTheme(<SearchMyFlightDetailsForm />);
    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  describe('when flight details received', () => {
    test('renders successfully', () => {
      // Arange
      (useMyFlightsContext as unknown as jest.Mock).mockImplementation(
        jest.fn(() => ({
          flightsResult: { chosenFlight: mockedDeserializedFlights[0] },
        }))
      );
      // Act
      const { baseElement } = renderWithTheme(<SearchMyFlightDetailsForm />);
      // Assert
      expect(baseElement).toMatchSnapshot();
    });
  });
});
