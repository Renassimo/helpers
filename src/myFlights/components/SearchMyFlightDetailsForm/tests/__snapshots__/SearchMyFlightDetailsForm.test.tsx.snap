// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchMyFlightDetailsForm renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div>
        MockedFlightForm
      </div>
      <div>
        MockedAirportForm - 
        Origin
      </div>
      <div>
        MockedAirportForm - 
        Destination
      </div>
      <div>
        MockedAircraftForm - 
      </div>
    </div>
  </div>
</body>
`;

exports[`SearchMyFlightDetailsForm when flight details received renders successfully 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div>
        MockedFlightForm
      </div>
      <div>
        MockedAircraftForm - 
        A6-FMB
      </div>
    </div>
  </div>
</body>
`;
